doc/operations/code_upgrade.rst

```diff
@@ -0,0 +1,191 @@
+Code Upgrade
+============
+
+.. note:: Refer to :doc:`/operations/code_upgrade_1x` for specific guidelines on how to upgrade a 1.x CCF service to 2.0.
+
+This page describes how operators/members can upgrade a live CCF service to a new version with minimal downtime.
+
+Reasons for running the code upgrade procedure include:
+
+- Upgrading nodes to a new version of a C++ application or JavaScript runtime (i.e. ``libjs_generic.enclave.so.signed``).
+- Upgrading nodes to a new version of CCF.
+
+.. tip::
+
+    - Note that there is no need to run the code upgrade procedure detailed on this page if `only` the JavaScript/TypeScript application needs updating (see :ref:`JavaScript/TypeScript bundle deployment procedure <build_apps/js_app_bundle:Deployment>`).
+    - If more than a majority of nodes have failed, the disaster recovery procedure should be run by operators instead (see :doc:`/operations/recovery`).
+
+.. note:: CCF guarantees specific live compatibility across different LTS versions. See :ref:`overview/release_policy:Operations compatibility` for more details.
+
+Procedure
+---------
```

Comment:
The code upgrade procedure described here is very much the slowest possible one (add new nodes one by one, retired old ones one by one). We know we should be able to support an atomic proposal to retire all old nodes and add the new ones in one go. However, we don't have a test for this so recommending the step-by-step approach seems like the most sensible option for now.

Reply:
There's an intermediate variant where nodes are added as a batch, and removed as a batch but in separate transactions. Agree that we shouldn't describe something we don't have a CI test for.

Reply:
Yes, good point. I've created https://github.com/microsoft/CCF/issues/3390 to cover this in the future.